bug: socket_recv_timeout_test.sfn flakily aborts (SIGABRT/exit 134) in the linux e2e-b shard

## Goal
Root-cause and fix the intermittent abort of `compiler/tests/e2e/socket_recv_timeout_test.sfn`, which fails CI non-deterministically on the `linux-x86_64 / e2e-b` shard and blocks unrelated PRs from going green.

## Symptom
The e2e-b shard fails with:

```
═══ socket_recv_timeout_test.sfn: 0/1 passed, 1 failed ═══
  socket_recv_timeout_test.sfn
make: *** [Makefile:415: test-shard] Error 134
##[error]Process completed with exit code 2.
```

Exit **134 = 128 + 6 = SIGABRT** — the test process *aborts*, it does not merely time out. Every other test in the shard passes; the failure is isolated to this one test and does not reproduce on every run.

## Frequency / recency
Observed **multiple times today** (2026-07-03) and reported to have **started recently** — a regression signal, not a long-standing flake. Most recent instance: PR #1861's run [28668337831, job 85025948431](https://github.com/SailfinIO/sailfin/actions/runs/28668337831/job/85025948431). That PR's diff is comments + one markdown file only, so the failure is unrelated to the PR under test — it is a pre-existing intermittent abort surfacing on an innocent branch.

## What the test does
`socket_recv_timeout_test.sfn` (from #1581, gap B3 of epic #1540) mirrors the `SO_RCVTIMEO`/`SO_SNDTIMEO` mechanism over a raw libc loopback socket pair (fixed port **18793**), asserting (A) a recv against a silent peer returns `EAGAIN` within the timeout, and (B) a normal recv still succeeds. It calls libc `socket`/`bind`/`listen`/`accept`/`connect`/`send`/`recv`/`setsockopt` directly via `extern fn`, with `malloc`/`free`/`memset` for the `sockaddr`/`timeval` buffers.

Because it aborts rather than hangs, the likely fault is **not** the "case A blocks forever" path but something raising SIGABRT — candidates to investigate:
- A memory/pointer fault in the raw `malloc`/`memset`/pointer-offset buffer handling (`_st_ptr_off`, the `sockaddr`/`timeval` construction) triggering a glibc abort (e.g. detected corruption / double free) under load.
- A **fixed-port collision** on 18793 under the parallel `--jobs N` pool (bind/connect race) leading to an unexpected state the test's error handling aborts on.
- Interaction with the recent runtime/TLS wave that touched the socket/serve paths and process init — plausible "recently started" suspects: **#1848** (TLS termination in serve accept loop), **#1857** (system CA trust-store fallback), **#1814** (link OpenSSL + tls extern surface). Linking OpenSSL globally can change process/signal init even for tests that don't use TLS.

## Investigation pointers
- Reproduce under the parallel pool: `sailfin test compiler/tests/e2e --jobs N` (N>1) in a loop; a single-job run may not surface it.
- Capture the abort: run the built test binary under `gdb`/`catchsegv` or with `ASAN` (note: sanitizers need `SAILFIN_MEM_LIMIT=unlimited` + an uncapped shell per `.claude/rules/compiler-safety.md`) to get the aborting frame.
- Bisect against the recent runtime/TLS merges if a local repro is found.

## Acceptance Criteria
- [ ] Root cause identified (aborting frame / race / regression commit).
- [ ] `socket_recv_timeout_test.sfn` passes deterministically under the parallel `e2e` pool across repeated runs (e.g. 50+ consecutive `--jobs N` iterations green).
- [ ] `make check` green.

## Files Affected (advisory map — non-binding, expected to drift)
- `compiler/tests/e2e/socket_recv_timeout_test.sfn` — the test itself (buffer handling, port selection, error paths).
- `runtime/sfn/adapters/http.sfn`, `runtime/sfn/concurrency/serve.sfn` — the shipped `SO_*TIMEO` code the test mirrors (if the fault is in the real path).
- Recent suspects: the TLS/runtime changes in #1848 / #1857 / #1814.

## Type
Bug

## Size
S

## Relationships
- Surfaced on PR #1861 (unrelated comments-only change). Test origin: #1581 (epic #1540, native socket recv/send timeouts).

--- a/compiler/tests/e2e/socket_recv_timeout_test.sfn
+++ b/compiler/tests/e2e/socket_recv_timeout_test.sfn
@@ -51,14 +51,37 @@ extern fn close(fd: i32) -> i32;
 
 extern fn setsockopt(sockfd: i32, level: i32, optname: i32, optval: * u8, optlen: i32) -> i32;
 
-// Fixed loopback port for this test. High + uncommon, distinct from
-// `serve_loopback_test.sfn`'s 18790 so the two cannot collide under
-// `--jobs N`.
-fn _st_port() -> i64 { return 18793; }
+// Fixed loopback port for this test. High + uncommon, and — critically —
+// distinct from every other server-binding e2e test so the parallel
+// `--jobs N` pool cannot collide on it: 18790 serve_loopback, 18791
+// serve_reuseaddr, 18792 serve_keepalive, 18793 the long-lived ASAN-serve
+// server (`concurrency_ownedbuf_asan_interleave_test.sfn`), 18795+ the TLS
+// tests. This test previously shared 18793 with that ASAN-serve test; under
+// the pool its `serve()` server held 18793 for its ~60s lifetime, so this
+// test's raw `bind` lost the race with EADDRINUSE and `assert lfd >= 0`
+// aborted (SIGABRT / exit 134, #1862). 18794 is unshared.
+fn _st_port() -> i64 { return 18794; }
 
 fn _st_ptr_off(p: * u8, off: i64) -> * u8 {
     let addr: i64 = (p as i64) + off;
     return addr as * u8;
+}
+
+// Arm SO_REUSEADDR on the listening socket before `bind`, mirroring the
+// shipped `serve.sfn::_serve_set_reuseaddr` (#1583). The port de-collision
+// above is the primary #1862 fix; this hardens the raw `bind` against a
+// lingering TIME_WAIT socket on a repeated run (e.g. the `make check`
+// triple self-host pass rebinding inside the 60s TIME_WAIT window). 4-byte
+// int optval = 1; both platforms armed (Linux SOL_SOCKET=1/SO_REUSEADDR=2,
+// macOS 65535/4), the wrong-platform call failing harmlessly.
+fn _st_set_reuseaddr(fd: i32) -> void {
+    let optval: * u8 = malloc(4 as usize);
+    if optval as i64 == 0 { return; }
+    memset(optval, 0, 4 as usize);
+    memset(_st_ptr_off(optval, 0), 1, 1 as usize);
+    setsockopt(fd, 1, 2, optval, 4);
+    setsockopt(fd, 65535, 4, optval, 4);
+    free(optval);
 }
 
 // Open a stream socket bound to `0.0.0.0:port` and listen. Tries both
@@ -71,6 +94,7 @@ fn _st_bind_listen(port: i64) -> i32 {
         if attempt >= 2 { break; }
         let fd: i32 = socket(2, 1, 0);
         if fd < 0 { return 0 - 1; }
+        _st_set_reuseaddr(fd);
         let addr: * u8 = malloc(16 as usize);
         if addr as i64 == 0 {
             close(fd);

